### ISSUE TITLE
CCN Histogram issue - Duplicates in feature file

### ISSUE THREAD

**Issue by bgrundy** [simsong/bulk_extractor#90]
I had an actual case where I could not resolve the CCN histogram with results of file analysis on the disk.  Turns out the CCN feature file had duplicate entries that were skewing the histogram.  I'm not sure if I'm missing a simple explanation for this, but I created a simple test file with data from a fake cc generator to test, and the results were the same.  The only thing I can think of is multiple scanners running over the data and reporting some of the results more than once.  But I can't find proof of this in the documentation.

```
$ bulk_extractor -V
bulk_extractor 1.5.5

$ cat CCTest.txt 
http://www.getcreditcardnumbers.com/
4229947005705149
4929603383331767
4024007145695358
4539882460939666
4817948058189932
5541719225054089
5573478997920404
5179261220551321
5331204135059639
5409087673501717
6011440279041729
6011784302359032
6011228165834913
6011264412856466
6011782281824877

$ bulk_extractor -o bcctest CCTest.txt

$ cat bcctest/ccn.txt 
# BULK_EXTRACTOR-Version: 1.5.5 ($Rev: 10844 $)
# Feature-Recorder: ccn
# Filename: CCTest.txt
# Feature-File-Version: 1.1
17	4929603383331767	229947005705149\x0A4929603383331767\x0A402400714569535
34	4024007145695358	929603383331767\x0A4024007145695358\x0A453988246093966
51	4539882460939666	024007145695358\x0A4539882460939666\x0A481794805818993
68	4817948058189932	539882460939666\x0A4817948058189932\x0A554171922505408
85	5541719225054089	817948058189932\x0A5541719225054089\x0A557347899792040
102	5573478997920404	541719225054089\x0A5573478997920404\x0A517926122055132
119	5179261220551321	573478997920404\x0A5179261220551321\x0A533120413505963
136	5331204135059639	179261220551321\x0A5331204135059639\x0A540908767350171
153	5409087673501717	331204135059639\x0A5409087673501717\x0A601144027904172
170	6011440279041729	409087673501717\x0A6011440279041729\x0A601178430235903
187	6011784302359032	011440279041729\x0A6011784302359032\x0A601122816583491
204	6011228165834913	011784302359032\x0A6011228165834913\x0A601126441285646
221	6011264412856466	011228165834913\x0A6011264412856466\x0A601178228182487
238	6011782281824877	011264412856466\x0A6011782281824877\x0A
17	4929603383331767	229947005705149\x0A4929603383331767\x0A402400714569535
51	4539882460939666	024007145695358\x0A4539882460939666\x0A481794805818993
85	5541719225054089	817948058189932\x0A5541719225054089\x0A557347899792040
119	5179261220551321	573478997920404\x0A5179261220551321\x0A533120413505963
153	5409087673501717	331204135059639\x0A5409087673501717\x0A601144027904172
187	6011784302359032	011440279041729\x0A6011784302359032\x0A601122816583491
221	6011264412856466	011228165834913\x0A6011264412856466\x0A601178228182487

$ cat bcctest/ccn.txt | grep ^51
51	4539882460939666	024007145695358\x0A4539882460939666\x0A481794805818993
51	4539882460939666	024007145695358\x0A4539882460939666\x0A481794805818993

$ cat bcctest/ccn_histogram.txt        
# BULK_EXTRACTOR-Version: 1.5.5 ($Rev: 10844 $)
# Feature-Recorder: ccn
# Filename: CCTest.txt
# Histogram-File-Version: 1.1
n=2	4539882460939666
...

```
The above is with a simple text file with generated CC numbers.   On the real disk image it was much harder to find the duplicate entries among all the false positives (which I understand and appreciate).  Is there a reason I'm missing why there are duplicate entries for many of the CCN's (and at the same offsets)?  Several of the CCNs were reported twice, but not all of them.

Thanks.  BE is one of the few tools that I run on *every* case, and I much appreciate the work done on this.

**Comment by simsong** [comment]
Thanks for the clear and easy-to-read bug report.

I have not seen this behavior before. You have provided a very simple test case.

On which operating system are you running?

**Comment by bgrundy** [comment]
$ uname -a
Linux hermes 4.4.19 #2 SMP Mon Aug 22 14:37:39 CDT 2016 x86_64 Intel(R) Core(TM) i5-3550 CPU @ 3.30GHz GenuineIntel GNU/Linux

This is a Slackware64 14.2 install with bulk_extractor built from https://www.slackbuilds.org/repository/14.2/system/bulk_extractor/

**Comment by simsong** [comment]
I am not able to duplicate the problem.

```
[Dance ~/gits/bulk_extractor/tests 00:46:12](master) $ uname -a
Darwin Dance.local 16.4.0 Darwin Kernel Version 16.4.0: Thu Dec 22 22:53:21 PST 2016; root:xnu-3789.41.3~3/RELEASE_X86_64 x86_64
[Dance ~/gits/bulk_extractor/tests 00:46:02](master) $ cat CCTest.txt
https://github.com/simsong/bulk_extractor/issues/90
http://www.getcreditcardnumbers.com/
4229947005705149
4929603383331767
4024007145695358
4539882460939666
4817948058189932
5541719225054089
5573478997920404
5179261220551321
5331204135059639
5409087673501717
6011440279041729
6011784302359032
6011228165834913
6011264412856466
6011782281824877
[Dance ~/gits/bulk_extractor/tests 00:44:28](master) $ ../src/bulk_extractor -o bcctest CCTest.txt
bulk_extractor version: 1.6.0-dev
Hostname: Dance.local
Input file: CCTest.txt
Output directory: bcctest
Disk Size: 344
Threads: 4
Attempt to open CCTest.txt
All data are read; waiting for threads to finish...
Time elapsed waiting for 1 thread to finish:
     (timeout in 60 min.)
All Threads Finished!
Producer time spent waiting: 0 sec.
Average consumer time spent waiting: 0.075674 sec.
MD5 of Disk Image: 8d673a69397e25c09fea2a0d50323e6d
Phase 2. Shutting down scanners
Phase 3. Creating Histograms
Elapsed time: 0.117958 sec.
Total MB processed: 0
Overall performance: 0.00291629 MBytes/sec (0.000729073 MBytes/sec/thread)
Total email features found: 0
[Dance ~/gits/bulk_extractor/tests 00:44:41](master) $ more bcctest/ccn.txt
# BANNER FILE NOT PROVIDED (-b option)
# BULK_EXTRACTOR-Version: 1.6.0-dev ($Rev: 10844 $)
# Feature-Recorder: ccn
# Filename: CCTest.txt
# Feature-File-Version: 1.1
89      4229947005705149        ardnumbers.com/\x0A4229947005705149\x0A492960338333176
106     4929603383331767        229947005705149\x0A4929603383331767\x0A402400714569535
123     4024007145695358        929603383331767\x0A4024007145695358\x0A453988246093966
140     4539882460939666        024007145695358\x0A4539882460939666\x0A481794805818993
157     4817948058189932        539882460939666\x0A4817948058189932\x0A554171922505408
174     5541719225054089        817948058189932\x0A5541719225054089\x0A557347899792040
191     5573478997920404        541719225054089\x0A5573478997920404\x0A517926122055132
208     5179261220551321        573478997920404\x0A5179261220551321\x0A533120413505963
225     5331204135059639        179261220551321\x0A5331204135059639\x0A540908767350171
242     5409087673501717        331204135059639\x0A5409087673501717\x0A601144027904172
259     6011440279041729        409087673501717\x0A6011440279041729\x0A601178430235903
276     6011784302359032        011440279041729\x0A6011784302359032\x0A601122816583491
293     6011228165834913        011784302359032\x0A6011228165834913\x0A601126441285646
310     6011264412856466        011228165834913\x0A6011264412856466\x0A601178228182487
327     6011782281824877        011264412856466\x0A6011782281824877\x0A
[Dance ~/gits/bulk_extractor/tests 00:44:46](master) $ more bcctest/ccn
ccn.txt                   ccn_histogram.txt         ccn_track2.txt            ccn_track2_histogram.txt
[Dance ~/gits/bulk_extractor/tests 00:44:46](master) $ more bcctest/ccn_histogram.txt
# BANNER FILE NOT PROVIDED (-b option)
# BULK_EXTRACTOR-Version: 1.6.0-dev ($Rev: 10844 $)
# Feature-Recorder: ccn
# Filename: CCTest.txt
# Histogram-File-Version: 1.1
n=1     4024007145695358
n=1     4229947005705149
n=1     4539882460939666
n=1     4817948058189932
n=1     4929603383331767
n=1     5179261220551321
n=1     5331204135059639
n=1     5409087673501717
n=1     5541719225054089
n=1     5573478997920404
n=1     6011228165834913
n=1     6011264412856466
n=1     6011440279041729
n=1     6011782281824877
n=1     6011784302359032

```

**Comment by bgrundy** [comment]
Just to close this loop:  The culprit turns out to be lightgrep...I should have known to test on a clean configuration for bulk_extractor.  My apologies.

I tested using the 1.6.0-Dev this time (same test file and environment as previous):
[with "./configure --enable-lightgrep"]
```
cat bedev/ccn_histogram.txt 
# BANNER FILE NOT PROVIDED (-b option)
# BULK_EXTRACTOR-Version: 1.6.0-dev ($Rev: 10844 $)
# Feature-Recorder: ccn
# Filename: CCText.txt
# Histogram-File-Version: 1.1
n=2	4024007145695358
n=2	4229947005705149
n=2	4817948058189932
n=2	5331204135059639
n=2	5573478997920404
n=2	6011228165834913
n=2	6011440279041729
n=2	6011782281824877
n=1	4539882460939666
n=1	4929603383331767
n=1	5179261220551321
n=1	5409087673501717
n=1	5541719225054089
n=1	6011264412856466
n=1	6011784302359032
```
And now without including lightgrep:

```
cat bodev2/ccn_histogram.txt 
# BANNER FILE NOT PROVIDED (-b option)
# BULK_EXTRACTOR-Version: 1.6.0-dev ($Rev: 10844 $)
# Feature-Recorder: ccn
# Filename: CCText.txt
# Histogram-File-Version: 1.1
n=1	4024007145695358
n=1	4229947005705149
n=1	4539882460939666
n=1	4817948058189932
n=1	4929603383331767
n=1	5179261220551321
n=1	5331204135059639
n=1	5409087673501717
n=1	5541719225054089
n=1	5573478997920404
n=1	6011228165834913
n=1	6011264412856466
n=1	6011440279041729
n=1	6011782281824877
n=1	6011784302359032

```

**Comment by simsong** [comment]
I'm looking at this again. Did you have *both* the lightgreap and the conventional flex-based scanners running? If so, you may have had two scanners looking for the same detail.

**Comment by bgrundy** [comment]
Yes, that is exactly what was happening, and had I been mindful of the fact that I was compiling with liblightgrep support, I would have known to test that.  

There's still an issue with using lightgrep though (I'm guessing it's NOT a bulk_extractor issue though) in that it is missing some of the numbers the accts scanner is picking up.  

All the following examples were run using bulk_extractor with liblightgrep support compiled in (note that -E lightgrep results in a seg fault - guessing it's not a standalone scanner?:

```
$ bulk_extractor -o withaccts -E accts CCTest.txt 
bulk_extractor version: 1.5.5...

$ cat withaccts/ccn_histogram.txt 
# BANNER FILE NOT PROVIDED (-b option)
# BULK_EXTRACTOR-Version: 1.5.5 ($Rev: 10844 $)
# Feature-Recorder: ccn
# Filename: CCTest.txt
# Histogram-File-Version: 1.1
n=1	4024007145695358
n=1	4229947005705149
n=1	4539882460939666
n=1	4817948058189932
n=1	4929603383331767
n=1	5179261220551321
n=1	5331204135059639
n=1	5409087673501717
n=1	5541719225054089
n=1	5573478997920404
n=1	6011228165834913
n=1	6011264412856466
n=1	6011440279041729
n=1	6011782281824877
n=1	6011784302359032

$ bulk_extractor -o withlg -E lightgrep  CCTest.txt 
Segmentation fault

$ bulk_extractor -o withlg -x all -e lightgrep  CCTest.txt 
Segmentation fault

$ bulk_extractor -o withlg2 -x accts -e lightgrep  CCTest.txt 
bulk_extractor version: 1.5.5...

$ cat withlg2/ccn_histogram.txt 
# BANNER FILE NOT PROVIDED (-b option)
# BULK_EXTRACTOR-Version: 1.5.5 ($Rev: 10844 $)
# Feature-Recorder: ccn
# Filename: CCTest.txt
# Histogram-File-Version: 1.1
n=1	4024007145695358
n=1	4229947005705149
n=1	4817948058189932
n=1	5331204135059639
n=1	5573478997920404
n=1	6011228165834913
n=1	6011440279041729
n=1	6011782281824877
```
BUT, if accts_lg is disabled, then even with lightgrep enabled all the numbers are found without duplicate.  I'm having a hard time deciphering this from the documentation.

```
$ bulk_extractor -o wo_lg  -e lightgrep -x accts_lg CCTest.txt 
bulk_extractor version: 1.5.5...

$ cat wo_lg/ccn_histogram.txt 
# BANNER FILE NOT PROVIDED (-b option)
# BULK_EXTRACTOR-Version: 1.5.5 ($Rev: 10844 $)
# Feature-Recorder: ccn
# Filename: CCTest.txt
# Histogram-File-Version: 1.1
n=1	4024007145695358
n=1	4229947005705149
n=1	4539882460939666
n=1	4817948058189932
n=1	4929603383331767
n=1	5179261220551321
n=1	5331204135059639
n=1	5409087673501717
n=1	5541719225054089
n=1	5573478997920404
n=1	6011228165834913
n=1	6011264412856466
n=1	6011440279041729
n=1	6011782281824877
n=1	6011784302359032

```
Bottom line is that my original report was based on running multiple scanners on the data and I missed that.  But at the very least there is a problem when using lightgrep and relying on the output where CC data is of interest.  Running with lightgrep can result in missed legitimate account hits (with -x accts)  OR duplicate hits (without -x accts).  

Thanks for your time and patience.  As this does not appear to be a BE issue, I'll continue testing to see if I can figure out what's happening.

**Comment by simsong** [comment]
You have discovered that lightgrep's scanner does not perfectly replace the bulk_extractor scanner, despite the fact that is what they set out to do.  You could look at at the source code of the two scanners and migrate over the functionality that is still missing...

**Comment by bgrundy** [comment]
Alas, the proverbial "infinite monkeys" would stand a better chance...but I will certainly look to it as an educational exercise.  I appreciate the response, and your time.
